File: Java-Sample-App/app/src/main/java/com/payu/payuppisdksample/MainActivity.java
package com.payu.payuppisdksample;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;

import com.payu.onepayujssdk.OnePayUJSSDK;
import com.payu.onepayujssdk.listeners.OnePayUJSHashGenerationListener;
import com.payu.onepayujssdk.listeners.OnePayUJSListener;
import com.payu.onepayujssdk.models.OnePayUJSHashes;
import com.payu.onepayujssdk.models.OnePayUJSParams;

import java.security.MessageDigest;
import java.util.HashMap;

public class MainActivity extends AppCompatActivity implements OnePayUJSListener {

    private static final String TAG = "OnePayUJSSDK";
    String merchantKey = "smsplus";
    String salt = "1b1b0";
    OnePayUJSParams onePayUJSParams;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }


    public OnePayUJSParams prepareParams() {
        String requestId = System.currentTimeMillis() + "payu";
        onePayUJSParams = new OnePayUJSParams("PROD", merchantKey, "9528340384", "OLW", "1003097", requestId, "https://test.payu.in", null, null);
        return onePayUJSParams;
    }

    public void openCards(View view) {
        OnePayUJSSDK.showCards(this, prepareParams(), this);
    }

    private String calculateHash(String hashString) {
        Log.d(TAG, "calculateHash: " + hashString);
        try {
            MessageDigest messageDigest = MessageDigest.getInstance("SHA-512");
            messageDigest.update(hashString.getBytes());
            byte[] mdbytes = messageDigest.digest();
            return getHexString(mdbytes);
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    private String getHexString(byte[] array) {
        StringBuilder hash = new StringBuilder();
        for (byte hashByte : array) {
            hash.append(Integer.toString((hashByte & 0xff) + 0x100, 16).substring(1));
        }
        return hash.toString();
    }

    @Override
    public void generateHash(@NonNull HashMap<String, String> dataMap, @NonNull OnePayUJSHashGenerationListener onePayUJSHashGenerationListener) {
        String hashName = dataMap.get("hashName");
        String hashString = dataMap.get("hashString");
        Log.d(TAG, "generateHash: " + hashName + "-----------" + hashString);
        // Do not use this function to calculate hash, calculate hash on your server only
        String hash = calculateHash(hashString + salt);
        if (!TextUtils.isEmpty(hash)) {
            HashMap<String, String> hashMap = new HashMap();
            hashMap.put(hashName, hash);
            onePayUJSHashGenerationListener.onHashGenerated(hashMap);
        }
    }

    @Override
    public void onCancel() {
        Log.d(TAG, "onCancel: ");
    }

    @Override
    public void onError(int i, @NonNull String s) {
        Log.d(TAG, "onError: " + i + "-------------" + s);
    }
}

File: Kotlin-Sample-App/app/src/main/java/com/example/sampleapp/MainActivity.kt
package com.example.sampleapp

import android.app.Activity
import android.os.Bundle
import android.os.Environment
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.FragmentActivity
import com.payu.onepayujssdk.OnePayUJS
import com.payu.onepayujssdk.enums.EnvironmentType
import com.payu.onepayujssdk.listeners.OnePayUJSHashGenerationListener
import com.payu.onepayujssdk.listeners.OnePayUJSListener
import com.payu.onepayujssdk.models.OnePayUJSParams

class MainActivity : FragmentActivity(), OnePayUJSListener {

    //Test Key and Salt
    private val testKey = "smsplus"
    private val testSalt = "1b1b0"

    //Prod Key and Salt
    private val prodKey = "3TnMpV"
    private val prodSalt = "g0nGFe03"
    private var etEnvironment: EditText? = null
    private var etKey: EditText? = null
    private var etSalt: EditText? = null
    private var etRedirectUrl: EditText? = null
    private var etPhone: EditText? = null
    private var etWalletIdentifier: EditText? = null
    private var etUrn: EditText? = null
    private var etRequestID: EditText? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN)

//        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        etEnvironment = findViewById(R.id.etEnvironment)
        etKey = findViewById(R.id.etKey)
        etSalt = findViewById(R.id.etSalt)
        etRedirectUrl = findViewById(R.id.etRedirectUrl)
        etPhone = findViewById(R.id.etPhone)
        etWalletIdentifier = findViewById(R.id.etWalletIdentifier)
        etUrn = findViewById(R.id.etUrn)
        etRequestID = findViewById(R.id.etRequestID)
        setInitialData()

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
    }

    fun setInitialData(){
        etKey?.setText(testKey)
        etSalt?.setText(testSalt)
        etPhone?.setText("9528340384")
        etWalletIdentifier?.setText("OLW")
        etUrn?.setText("1003097")
    }

    fun openCards(view: View) {
        val onePayUJSParams = prepareParams()
        Log.i(
            "OnePayULog",
            "Data - onePayUJSParams: $onePayUJSParams"
        )

        OnePayUJS.showCards(this, onePayUJSParams, this)
    }

    override fun onCancel() {
        Log.i(
            "OnePayULog",
            "onCancel called"
        )
        Toast.makeText(this, "onCancel called", Toast.LENGTH_SHORT).show()
    }

    override fun onError(code: Int, message: String) {
        Log.i(
            "OnePayULog",
            "onError called - code: $code | message: $message"
        )
        Toast.makeText(this, "message: $message | $code", Toast.LENGTH_SHORT).show()
    }

    override fun generateHash(
        map: HashMap<String, String?>,
        hashGenerationOnePayUJKListener: OnePayUJSHashGenerationListener
    ) {
        if (map.containsKey(Constants.HASH_STRING) && map.containsKey(Constants.HASH_NAME)
        ) {
            val hashData = map[Constants.HASH_STRING]
            val hashName = map[Constants.HASH_NAME]
            println("OnePayULog hashName and hashData : $hashName | $hashData ")

            val hash: String? = HashGenerationUtils.generateHashFromSDK(
                hashData!!, etSalt?.text.toString()
            )

            if (!TextUtils.isEmpty(hash)) {
                val hashMap: HashMap<String, String> = HashMap()
                hashMap[hashName!!] = hash!!
                Log.i(
                    "OnePayULog",
                    "generateHash called hash: $hash"
                )
                hashGenerationOnePayUJKListener.onHashGenerated(hashMap)
            }
        }
    }

    private fun prepareParams(): OnePayUJSParams {
        val additionalParamsMap: HashMap<String, Any?> = HashMap()
        val requestId = System.currentTimeMillis().toString()+"payu"

        return OnePayUJSParams(
            environment = etEnvironment?.text.toString().uppercase(),
            merchantKey = etKey?.text.toString(),
            mobileNumber = etPhone?.text.toString(),
            walletIdentifier = etWalletIdentifier?.text.toString(),
            walletUrn = etUrn?.text.toString(),
            referenceId = requestId,
            additionalParams = additionalParamsMap
        )
    }
}